`RPC`-`Bridge` integration test

esanzgar opened the pull request:
We have agreed that `RPC` and `Bridge` classes would benefit from a more
realistic testing. These tests exercise `RPC` and `Bridge` code together
using `MessagePort`.

A second commit contains a small change in the current behaviour of `Bridge`:

Previously, executing the `call` method after the channels were
destroyed caused an asynchronous timeout and the promise to be
*resolved* with value of `[null, null...]`.

With this change, the `call` method returns immediately with an empty
array of results. This behaviour is similar to what happens trying to
use the `call` method before `createChannel`, like this:   https://github.com/hypothesis/client/blob/51f6d3ac02857e17c61665b08efa414a3f965b32/src/shared/test/integration/rpc-bridge-test.js#L158-L162

robertknight reviewed (commented):
Hi @esanzgar,

The main feedback I have here is that integration tests should try to avoid covering things that are effectively covered by unit tests and instead focus on the most important scenarios and failure modes that exercise the integration between the components. Since these integration tests are usually checking scenarios/activities rather than specific methods, the test descriptions will tend to reflect that. For example:

```js
it("should invoke Bridge `onConnect` callbacks after connecting") { ... }
it('should invoke Bridge method handlers when calling an RPC method', () => { ... });
it('should return the result of a successful call', () => { ... });
it('should return an error if an RPC call fails', () => { ... });
it('should close ports after destroying a Bridge', () => { ... });
```

It can be still be useful to group related tests together (eg. in "establishing a connection" or "calling RPC methods" blocks), and sometimes it will make sense to use the method/function that serves as the entry point to the scenario. If you're using a method, make sure it is clear what the class is though. In these tests all the `#methodName` describe blocks refer to methods of `Bridge`, but that may not be immediately obvious to a future reader who isn't as familiar with the code.

For this PR, I think what I would suggest to move forwards is to go through the tests, ask whether they are testing something that really involves integration between the components remove it if not. Then for the remaining tests revise the descriptions to either be more scenario focused or at least make sure it is clear which entry point (eg. `Bridge#call`) is being referred to.

esanzgar commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,277 @@
+  });
+
+  afterEach(() => {
+    bridges.forEach(bridge => bridge.destroy);
```
Good catch! Thanks!

esanzgar commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,277 @@
+  });
+
+  describe('#on', () => {
+    it('raise an error if trying to register a listener after a channel has been already created', () => {
```
I removed this test.

esanzgar commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,277 @@
+    it('returns an empty array if call method before a channel is created', async () => {
+      const bridge = createBridge();
+      const results = await bridge.call('method1', 'params1');
+      assert.equal(results.length, 0);
```
Correct, I moved this test to the `Bridge` unit testing.

esanzgar commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,277 @@
+    clock.restore();
+  });
+
+  describe('#createChannel', () => {
```
I agree, I have changed the descriptions and reorganised the tests.

esanzgar commented:
@robertknight, based on your feedback, I have:

* changed the language of the test descriptions and focus on
  activities or common scenarios

* grouped related tests together based on the different activities

* moved one unit test to `bridge-test.js` and delete tests that are covered
  in unit tests.

You can see these changes in the last commit.

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# Date:      Tue Jul 27 18:37:41 2021 +0200
#
# On branch rpc-bridge-integration-test
# Changes to be committed:
#	modified:   src/shared/test/bridge-test.js
#	modified:   src/shared/test/integration/rpc-bridge-test.js
#

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,206 @@
+    clock.restore();
+  });
+
+  context('establishing a connection', () => {
```
`describe` is more appropriate than `context` here. Functionally I believe they are exactly the same. However `context` is normally used to refer to a situation or circumstance (eg. "browser doesn't support feature X") whereas `describe` refers to _what_ is being tested, whether that is a specific class/function/method, as in most unit tests, or a process, as is the case here.

robertknight commented on src/shared/test/bridge-test.js:
```diff
@@ -99,6 +99,11 @@ describe('shared/bridge', () => {
 
   describe('#call', () => {
+    it('returns an empty array calling a RPC method before a channel is created', async () => {
+      const results = await bridge.call('method1', 'params1');
```
```suggestion
      const results = await bridge.call('method1', 'param1');
```

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,206 @@
+      assert.deepEqual(results.sort(), ['params1foo', 'params1bar'].sort());
+    });
+
+    it('should invoke Bridge method handler on every channel when calling a RPC method (with callback)', done => {
```
This is out of scope for this PR - but could we get rid of support for result callbacks in the `Bridge`? It looks to me like it duplicates the Promise return functionality and the Promise-based API is more convenient for consumers and easier to specify the type of.

esanzgar replied:
I am fine removing this test. It was just to demonstrated that the `error` is `null`, which is only possible to be tested in callback-based API (it is functionally the same as no error raised in the Promise-based).

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,206 @@
+
+      let error;
+      try {
+        await reciprocalBridge.call('method1', 'params1');
```
```suggestion
        await reciprocalBridge.call('method1', 'param1');
```

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,206 @@
+      assert.deepEqual(reciprocalBridge.links, []);
+    });
+
+    it('no longer send RPC messages to a Bridge channel that has received an error', async () => {
```
```suggestion
    it('no longer sends RPC messages to a Bridge channel that has received an error', async () => {
```

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,206 @@
+      const bridge = createBridge();
+      bridge.createChannel(port1);
+
+      bridge.call('method1', 'params1', (err, results) => {
```
I suggest to use the Promise-based API for collecting results in all these tests. This will make them easier to update if we remove the callback-based API in future.

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,206 @@
+
+      reciprocalBridge.destroy();
+
+      bridge.call('method2', 'params1', (err, results) => {
```
As with the previous test, if we can use the Promise-based API here this will make it easier to remove the callback API in future.

robertknight reviewed (commented):
The general changes look good. The main feedback I had is to do with the APIs that `Bridge` provides for returning results. It provides both Promise-based and callback-based responses. I can't see a good reason to have both. In that case it makes sense to me to prefer the Promise-based API. If you're agreed, then I suggest using the Promise-based API for all tests except the one that is explicitly testing the callback API.

esanzgar commented:
@robertknight , I changed the tests that used the callback-based API to test solely the `Promise`-based.

I also added a comment in the `Bridge.call` method to indicate the future intention to retire the callback-based API.

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,185 @@
+import Bridge from '../../bridge';
+
+describe('RPC-Bridge integration', () => {
+  const sandbox = sinon.createSandbox();
```
I'm not sure if we actually need to use a custom sandbox here. This would normally only be needed if you wanted to temporarily patch properties of global objects (eg. a `window` method) for the duration of a test.

esanzgar replied:
We don't use the sandbox at all 😄 .

Because this integration test derives from a unit test, I forgot to remove it.

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,185 @@
+      reciprocalBridge1.createChannel(port2);
+      reciprocalBridge2.createChannel(otherChannel.port2);
+
+      const results = await bridge.call('method1', 'param1');
```
```suggestion
      const results = await bridge.call('method1', 'param1');
      
```

Having blank lines before and after the "action" phase of a test with many lines makes it easier to see which code is the setup, which code is the action being tested and which code is the assertions.

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,185 @@
+      } catch (err) {
+        error = err;
+      }
+      assert.equal(error, errorMessage);
```
```suggestion

      assert.equal(error, errorMessage);
```

robertknight commented on src/shared/test/integration/rpc-bridge-test.js:
```diff
@@ -0,0 +1,185 @@
+      }
+      assert.equal(error.message, errorMessage);
+
+      for (let i = 0; i < 5; ++i) {
```
Having 5 iterations here seems a bit redundant. Surely one would be enough?

esanzgar replied:
I removed the loop.

robertknight reviewed (approved):
LGTM.

I have some general thoughts that don't affect this PR:

I think we probably want to take a step back soon and look at the overall responsibilities of the `RPC` and `Bridge` classes and consider whether we really do need both of them. If we do, the distinction between them should be made clearer. Right now it's not easy to understand that looking at the JSDoc for the `Bridge` class.

Currently the differences that I can see are:

1. RPC supports a single channel whereas Bridge supports multiple channels and allows channels to be added and removed at any point
2. RPC requires all the method handlers to be registered up-front, whereas Bridge allows them to be added incrementally as long as they are all added before connections are created
3. RPC returns results via a callback, whereas Bridge wraps that to return a promise

(1) is an important distinction, whereas (2) and (3) may just be a result of history.

esanzgar commented:
> I think we probably want to take a step back soon and look at the overall responsibilities of the RPC and Bridge classes and consider whether we really do need both of them. If we do, the distinction between them should be made clearer. Right now it's not easy to understand that looking at the JSDoc for the Bridge class.
> [...]
> (1) is an important distinction, whereas (2) and (3) may just be a result of history.

I agree with you.

I have been thinking that if (1) is the main distinction, then maybe renaming the `Bridge` class to `Broadcaster` could reflect better the difference.
